controller_manager/src/controller_manager.cpp

```diff
@@ -574,6 +574,23 @@ controller_interface::ControllerInterfaceBaseSharedPtr ControllerManager::load_c
   controller_spec.next_update_cycle_time = std::make_shared<rclcpp::Time>(
     0, 0, this->get_node_clock_interface()->get_clock()->get_clock_type());
 
+  // We have to fetch the params_file at the time of loading the controller, because this way we can
+  // load them at the creating of the LifeCycleNode and this helps in using the features such as
+  // read_only params, dynamic maps lists etc
+  // Now check if the params_file parameter exist
+  const std::string param_name = controller_name + ".params_file";
+  std::string params_file;
+
+  // Check if parameter has been declared
+  if (!has_parameter(param_name))
+  {
+    declare_parameter(param_name, rclcpp::ParameterType::PARAMETER_STRING);
```

Comment:
can we make that read_only? because it wouldn't get updated any time after?

Reply:
Nope, here it doesn't make sense. Because when a parameter is set as `read_only`, they are expected to already exist/loaded when you instantiate the controller_manager. If this is the case, the spawners cannot set this parameter

I could be wrong, but this is what I've understood from the documentation

Reply:
ok I see, makes sense.
afaik one can't change the behavior after the node was created, i.e., setting a parameter read-only after reading it once. Should we give at least a warning if a user changes this parameter? (set-parameter callback)

Reply:
no, I mean the parameter of the CM node `controller_name + ".params_file"`.

Reply:
but now it isn't read_only, so there is no output if someone changes it? (as it is a parameter of the CM node, not of the controller node)

Reply:
Sorry for the confusion, as it isn't read_only, if someone changes it, then you won't see any output at all, and the next time you read the parameter it is changed. However, this case won't happen as for it to happen, you need to unload the controller and load it again, if not, at this following condition the `load_controller` will fail, if called twice for the same controller 
https://github.com/ros-controls/ros2_control/blob/3309c6dbaf4ab5140d13908bedb6fb471c422498/controller_manager/src/controller_manager.cpp#L1290-L1301

So, simply no read_only parameter should be good for our usecases